zookeeper-api/src/test/java/org/apache/helix/zookeeper/impl/TestZooKeeperConnection.java

```diff
@@ -57,6 +59,40 @@ void testPersistWatcher() throws Exception {
     zkClient.close();
   }
 
+  @Test (dependsOnMethods = "testPersistWatcher")
+  void testRecursivePersistWatcherWithOneTimeWatcher() throws Exception {
+    // reset counter
+    get_count[0] = 0;
+    Watcher watcher1 = new PersistRecurWatcher();
+    ZkClient zkClient =   new org.apache.helix.zookeeper.impl.client.ZkClient(ZK_ADDR);
+    IZkConnection _zk = zkClient.getConnection();
+    String path="/testRecursivePersistWatcher";
+    _zk.create(path, null, ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+    // register a persist listener on a path, change the ZNode 100 times, create 100 child ZNode,
+    // and expecting 200 events
+    _zk.addWatch(path, watcher1, AddWatchMode.PERSISTENT_RECURSIVE);
+    for (int i=0; i<count; ++i) {
+      _zk.writeData(path, "datat".getBytes(), -1);
+      _zk.create(path+"/c1_" +i, null, ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+      _zk.create(path+"/c1_" +i + "/c2", null, ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+    }
+    Assert.assertTrue(countDownLatch2.await(50000, TimeUnit.MILLISECONDS));
+
+    // register a one time listener on the path. Count the total number of event.
+    // ZK will over write the persist watcher and only trigger event once for child and data change.
+    _zk.readData(path, null, true);
+    _zk.getChildren(path, true);
+    for (int i=0; i<200; ++i) {
+      _zk.writeData(path, ("datat"+i).getBytes(), -1);
+      _zk.create(path+"/c2_" +i, null, ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+    }
+    Assert.assertTrue(TestHelper.verify(() -> {
+      return (get_count[0] == 302);
```

Comment:
for my own understanding, why 302?

Reply:
good question!
When we register a persist recursive listener on a path and issue 300 changes in the tree, we get 300 events.
Register a one time listener on that path, then it overwrite the persist watcher and only trigger event once for child and data change.
-> 302 changes.